pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -939,31 +934,39 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(ctx context.C
 
 // assignDefaultStorageClass updates the claim storage class if there is any, the claim is updated to the API server.
 // Ignores claims that already have a storage class.
-// TODO: if resync is ever changed to a larger period, we might need to change how we set the default class on existing unbound claims
-func (ctrl *PersistentVolumeController) assignDefaultStorageClass(ctx context.Context, claim *v1.PersistentVolumeClaim) (bool, error) {
+func (ctrl *PersistentVolumeController) assignDefaultStorageClass(ctx context.Context, claim *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
 	logger := klog.FromContext(ctx)
 
 	if storagehelpers.GetPersistentVolumeClaimClass(claim) != "" {
-		return false, nil
+		return claim, nil
 	}
 
 	class, err := util.GetDefaultClass(ctrl.classLister)
 	if err != nil {
-		return false, err
+		return nil, err
```

Comment:
Returning `nil` does not seem safe since it's further used without nil pointer checks by functions like `storagehelpers.IsDelayBindingProvisioning(claim)`.

Reply:
@RomanBednar 

I found that there is a lot of code like this in the controller. When an error is returned, other return values do not return nil. I don't quite understand this approach. This is not in line with Go programming habits. Generally speaking, when When an error is returned, other return values, whether nil or not, are untrustworthy to the caller and should not be used. In the controller code, although no caller is using these untrustworthy values Trust the return value, but I still believe that non-nil values should not be returned when an error occurs. This will cause confusion for other developers as to when they can trust these return values.

Reply:
@carlory I see what you mean, not sure about the reasoning behind other functions here but for `assignDefaultStorageClass` the aim was to have a function that returns `(bool, error)`. The bool just informs the caller whether the object passed in was modified not, but this does not imply an error in either case - that's the confusing part I guess.

In other words:
- if caller of this function gets `false,nil` return values  -> everything is ok and object was not modified on purpose. In this case that can be failing to find default storage class - it's not an error and code can continue knowing PVC did not change
- if caller of this function gets `false,err` return values -> something broke and object was not modified, code should not continue (or log errors)